contracts/nft/contracts/evm/UniversalNFTCore.sol

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.26;
+
+import "@zetachain/protocol-contracts/contracts/evm/GatewayEVM.sol";
+import {RevertOptions} from "@zetachain/protocol-contracts/contracts/evm/GatewayEVM.sol";
+import "../shared/UniversalNFTEvents.sol";
+import {ERC721Upgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol";
+import {ERC721URIStorageUpgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721URIStorageUpgradeable.sol";
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+abstract contract UniversalNFTCore is
+    ERC721Upgradeable,
+    ERC721URIStorageUpgradeable,
+    OwnableUpgradeable,
+    UniversalNFTEvents
+{
+    GatewayEVM public gateway;
+    address public universal;
+    uint256 public gasLimitAmount;
+
+    error InvalidAddress();
+    error Unauthorized();
+    error InvalidGasLimit();
+    error GasTokenTransferFailed();
+
+    modifier onlyGateway() {
+        if (msg.sender != address(gateway)) revert Unauthorized();
+        _;
+    }
+
+    function setGasLimit(uint256 gas) external onlyOwner {
+        if (gas == 0) revert InvalidGasLimit();
+        gasLimitAmount = gas;
+    }
+
+    function setUniversal(address contractAddress) external onlyOwner {
+        if (contractAddress == address(0)) revert InvalidAddress();
+        universal = contractAddress;
+        emit SetUniversal(contractAddress);
+    }
+
+    function __UniversalNFTCore_init(
+        address gatewayAddress,
+        address universalAddress,
+        uint256 gas
+    ) internal {
+        if (gatewayAddress == address(0)) revert InvalidAddress();
+        if (universalAddress == address(0)) revert InvalidAddress();
+        if (gas == 0) revert InvalidGasLimit();
+        gateway = GatewayEVM(gatewayAddress);
+        universal = universalAddress;
+        gasLimitAmount = gas;
+    }
+
+    /**
+     * @notice Transfers an NFT to another chain.
+     * @dev    Burns the NFT locally, then sends an encoded message to the
+     *         Gateway to recreate it on the destination chain (or revert if needed).
+     * @param  tokenId The ID of the NFT to transfer.
+     * @param  receiver The address on the destination chain that will receive the NFT.
+     * @param  destination The contract address on the destination chain (or address(0) if same chain).
```

Comment:
I think we could add a bit more explanation on the difference

Reply:
https://github.com/zeta-chain/standard-contracts/commit/151e9c39f69999180302ce93ea67c2668c7e8565